xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,363 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file contains implementation of a simple buffer that can store
+// data from multiple transfers over the XLS channel and glue them together.
+// It contains various operations that can simplify access to the stored data.
+//
+// Utility functions containing the `_checked` suffix, in addition
+// to performing the actual operation, return information whether the operation
+// was successful. If the precondition is guaranteed to be true, one can use
+// the function with the same name but without the suffix.
+
+import std
+
+// Structure to hold the buffered data
+pub struct Buffer<BSIZE: u32> {
+    contents: bits[BSIZE],
```

Comment:
I'm curious if generic bit storage will be more convenient that something with concrete types that related to the zstd algorithm, maybe you could add a section around this in the [design discussion](https://docs.google.com/document/u/0/d/1wYtfG-IoXZL0AmpwozYr0GE_txsEB4bpAaPwvgkPZ6o/edit)?

Reply:
This simple structure is just for buffering the input data. For example, we use it to aggregate enough data to parse the entire frame header. The extracted information will then be stored in the zstd decoder state.